File: YJ/조이스틱.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42860

def solution(name):
    answer = 0

    min_move = len(name) - 1
    for i, s in enumerate(name):
        answer += min(ord(s) - ord("A"), 26 - (ord(s) - ord("A")))

        t = i + 1
        while t < len(name) and name[t] == "A":
            t += 1
        
        min_move = min(min_move, 2 * i + (len(name) - t), 2 * (len(name) - t) + i)

    answer += min_move
    return answer

print(solution("BABBAABB")) # 11
print(solution("BAABBAAA")) # 7
# 아예 긴 거를 선택하는 거도 있음 => 그래서 그냥 0으로 된 뭉치를 한 개 배제하고 싹다 돌리자.

print(solution("JEROEN")) # 56
print(solution("JAN")) # 23
print(solution("JAZ")) # 11
print(solution("ABABBA")) # 7
print(solution("AAABBA")) # 5
print(solution("BBBAAAAB")) # 8
print(solution("BAAAAABAA")) # 5
print(solution("BBBABAABABABB")) # 20
print(solution("BBABAAAAAAB")) # 9
print(solution("BBAAAAAAABAB")) # 9

File: JK/조이스틱.js
function solution(name) {
  // 이동 횟수 + 알파벳
  let answer = 0;
  // 이동 횟수(최대 길이는 무조건 name의 길이 - 1)
  let move = name.length - 1;

  const arr = name.split("");
  arr.map((item, idx) => {
    // 정방향과 역방향의 코스트 비교
    answer += Math.min(item.charCodeAt() - 65, 26 - (item.charCodeAt() - 65));

    // A가 아닌 다음 문자
    let nextChar = idx + 1;

    while (nextChar < arr.length && arr[nextChar] === "A") {
      nextChar += 1;
    }

    // 역방향과 정방향 이동 비교
    // idx * 2 + arr.length - nextChar : 정방향 이동 -> 역방향 이동으로 target까지 도달
    // (arr.length - nextChar) * 2 + idx : 역방향 이동 -> 정방향 이동으로 target까지 도달
    move = Math.min(move, idx * 2 + arr.length - nextChar, (arr.length - nextChar) * 2 + idx);
  });
  answer += move;

  return answer;
}

console.log(solution("JEROEN"));
console.log(solution("JAN"));
